File: djikstra.cpp
#include <climits>
#include <functional>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> dijkstra(int V, vector<vector<int>> &edges, int src) {

    // Step 1: Build adjacency list
    vector<vector<pair<int, int>>> adj(V);

    for (auto &edge : edges) {
      int u = edge[0];
      int v = edge[1];
      int w = edge[2];

      adj[u].push_back({v, w});
      adj[v].push_back({u, w}); // remove this line if graph is directed
    }

    // Step 2: Min-heap (distance, node)
    priority_queue<pair<int, int>, vector<pair<int, int>>,
                   greater<pair<int, int>>>
        pq;

    // Step 3: Distance array
    vector<int> dist(V, INT_MAX);
    dist[src] = 0;

    pq.push({0, src});

    // Step 4: Dijkstra
    while (!pq.empty()) {
      int dis = pq.top().first;
      int node = pq.top().second;
      pq.pop();

      // Skip if we already found a better distance
      if (dis > dist[node])
        continue;

      for (auto &it : adj[node]) {
        int adjNode = it.first;
        int edgeWeight = it.second;

        if (dis + edgeWeight < dist[adjNode]) {
          dist[adjNode] = dis + edgeWeight;
          pq.push({dist[adjNode], adjNode});
        }
      }
    }

    return dist;
  }
};

int main() {
  int V = 5;

  vector<vector<int>> edges = {{0, 1, 2}, {0, 2, 4}, {1, 2, 1},
                               {1, 3, 7}, {2, 4, 3}, {3, 4, 1}};

  int src = 0;

  Solution sol;
  vector<int> result = sol.dijkstra(V, edges, src);

  cout << "Shortest distances from source " << src << ":\n";
  for (int i = 0; i < V; i++) {
    cout << "Node " << i << " -> ";
    if (result[i] == INT_MAX)
      cout << "INF";
    else
      cout << result[i];
    cout << endl;
  }

  return 0;
}

File: djikstra.py
from typing import List, Dict
import heapq


class Solution:
    def shortestPath(self, n: int, edges: List[List[int]], src: int) -> Dict[int, int]:
        # Build adjacency list
        adj = {i: [] for i in range(n)}
        for s, d, weight in edges:
            adj[s].append((d, weight))

        shortest = {}  # vertex -> shortest distance

        minHeap = [(0, src)]  # (distance, node)

        while minHeap:
            w1, n1 = heapq.heappop(minHeap)

            if n1 in shortest:
                continue

            shortest[n1] = w1

            for n2, w2 in adj[n1]:
                if n2 not in shortest:
                    heapq.heappush(minHeap, (w1 + w2, n2))

        # Mark unreachable nodes as -1
        for i in range(n):
            if i not in shortest:
                shortest[i] = -1

        return shortest


def main():
    n = 5
    edges = [[0, 1, 2], [0, 2, 4], [1, 2, 1], [1, 3, 7], [2, 4, 3], [3, 4, 1]]
    src = 0

    sol = Solution()
    result = sol.shortestPath(n, edges, src)

    print("Shortest distances from source:", src)
    for node in sorted(result.keys()):
        print(f"Node {node} -> {result[node]}")


if __name__ == "__main__":
    main()
